### PR TITLE
Add manual feature exposure recording

### PR DESCRIPTION
**Pull request by travis79**

### CODE REVIEW

File: components/nimbus/ios/Nimbus/Nimbus.swift
Comment (jhugman) [inline components/nimbus/ios/Nimbus/Nimbus.swift:53]: It took me a minute to persuade myself that this worked.

My reasoning: 

1. `getActiveExperiments` gets a list of experiments that are enrolled.
2. Features are only enrolled in zero or one experiment at a time.
3. The first experiment we find where the `featureIds` contains our `featureId` will therefore be the only experiment enrolled involving this experiment.

For the purposes of this PR: let's add a comment persuading the reader that this works.

Moving forward: Adding Rust FFI surface area in the same PR should be fine — previously it wasn't. We should probably keep this, for the sake of shipping, but start the habit of treating the FFI as internals, not an API itself.

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt
Comment (jhugman) [inline components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt:169]: This PR doesn't seem to call this method from `getVariables`. We should probably fix that before landing.

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt
Comment (jhugman) [inline components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt:165]: e.g. experimenting on a menu, where menu construction happens at a different time to the user seeing the menu. 

Or `getVariables` is being asked for multiple times, with the same feature, and we want to record exposure only once.

e.g. when `getVariables` is being asked for every keystroke.

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt
Comment (jhugman) [inline components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt:549]: This means:
1. `getExperimentBranch` by experiment slug is now not recording exposure events.
2. `getExperimentBranch` by feature id did not record exposure events prior to this change?

**Review by jhugman — COMMENTED** [review]
A few comments. I think we should wait until the swift variables land first.

File: components/nimbus/ios/Nimbus/Nimbus.swift
Comment (travis79) [reply components/nimbus/ios/Nimbus/Nimbus.swift:53]: I foresee where this entire function will be moved to Rust and can just be called across the FFI and recorded in Glean there (once I can get to it). So, rather than touching the Rust now and potentially making even more surface area that might need refactored when that happened, it seemed useful for me to try and keep this isolated in the foreign language bindings where the Glean integration is currently.

I agree definitely about adding a comment to clarify what exactly we are expecting from this line.

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt
Comment (travis79) [reply components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt:169]: Maybe a merge snafu, I will investigate.

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt
Comment (travis79) [reply components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/Nimbus.kt:549]: Correct. My thinking was that get branch by experiment slug was deprecated and this would help to support that. yes, we may have been missing exposure events for a couple of experiments since it was trying to match feature ids against the slug.

File: megazords/ios/MozillaAppServicesTests/NimbusTests.swift
Comment (jhugman) [inline megazords/ios/MozillaAppServicesTests/NimbusTests.swift:307]: Should this be accompanied with a negative test, called before the positive test?

**Review by jhugman — APPROVED** [review]
A couple of additional changes needed, but doesn't need to be held up by another review cycle. Thanks for jumping on this one!

File: components/nimbus/ios/Nimbus/Nimbus.swift
Comment (travis79) [reply components/nimbus/ios/Nimbus/Nimbus.swift:208]: I'd be happy to do that, actually.  The only thing that worries me is that it dilutes down the `GleanIntegration` piece if part of the integration will be defined in another extension. But, let me tinker with it and see if we can simplify things a bit.

Comment (travis79) [reply components/nimbus/ios/Nimbus/Nimbus.swift]: Per our convo on Slack, I was able to satisfy this by refactoring a bit so that everything Glean still lives inside the same extension definition.

**Comment by travis79** [comment]
Change requests addressed, awaiting checks to pass

**Pull request merged**
